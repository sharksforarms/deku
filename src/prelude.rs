pub use crate::{error::DekuError, BitsReader, BitsWriter, DekuRead, DekuWrite};
pub use bitvec::{
    order::BitOrder, order::Lsb0, order::Msb0, slice::AsBits, slice::BitSlice, vec::BitVec,
};
