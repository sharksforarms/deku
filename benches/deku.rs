use criterion::{black_box, criterion_group, criterion_main, Criterion};
use deku::prelude::*;

#[derive(Debug, PartialEq, DekuRead)]
struct DekuByte {
    data: u8
}

#[derive(Debug, PartialEq, DekuRead)]
#[deku(id_type = "u8")]
enum DekuEnum {
    #[deku(id = "0x01")]
    VariantA(u8),
}

#[derive(Debug, PartialEq, DekuRead)]
struct DekuVec {
    count: u8,
    #[deku(len = "count")]
    data: Vec<u8>
}

fn deku_read_byte(input: &[u8]) {
    let (_rest, v) = DekuByte::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuByte { data: 0x01 }, v);
}

fn deku_read_enum(input: &[u8]) {
    let (_rest, v) = DekuEnum::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuEnum::VariantA(0x02), v);
}

fn deku_read_vec(input: &[u8]) {
    let (_rest, v) = DekuVec::from_bytes((input, 0)).unwrap();
    assert_eq!(100, v.count);
    assert_eq!(100, v.data.len());
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("deku_read_byte", |b| {
        b.iter(|| deku_read_byte(black_box([0x01].as_ref())))
    });

    c.bench_function("deku_read_enum", |b| {
        b.iter(|| deku_read_enum(black_box([0x01, 0x02].as_ref())))
    });

    let deku_read_vec_input = {
        let mut v = [0xFFu8; 101].to_vec();
        v[0] = 100u8;
        v
    };
    c.bench_function("deku_read_vec", |b| {
        b.iter(|| deku_read_vec(black_box(&deku_read_vec_input)))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
