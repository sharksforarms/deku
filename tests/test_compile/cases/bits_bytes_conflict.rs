use deku::prelude::*;

#[derive(DekuRead)]
#[deku(id_type = "u8", bits = 1, bytes = 2)]
enum Test1 {}

#[derive(DekuRead)]
#[deku(id_type = "u8")]
enum Test2 {
    A(#[deku(bits = 1, bytes = 2)] u8),
    B {
        #[deku(bits = 3, bytes = 4)]
        a: u8,
    },
}

#[derive(DekuRead)]
struct Test3 {
    #[deku(bits = 5, bytes = 6)]
    a: u8,
}

#[derive(DekuRead)]
struct Test4(#[deku(bits = 7, bytes = 8)] u8);

fn main() {}
