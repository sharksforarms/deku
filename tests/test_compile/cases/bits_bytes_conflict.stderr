error: conflicting: both `bits` and `bytes` specified on enum
 --> $DIR/bits_bytes_conflict.rs:4:28
  |
4 | #[deku(type = "u8", bits = "1", bytes = "2")]
  |                            ^^^

error: conflicting: both `bits` and `bytes` specified on field
  --> $DIR/bits_bytes_conflict.rs:10:21
   |
10 |     A(#[deku(bits = "1", bytes = "2")] u8),
   |                     ^^^

error: conflicting: both `bits` and `bytes` specified on field
  --> $DIR/bits_bytes_conflict.rs:19:19
   |
19 |     #[deku(bits = "5", bytes = "6")]
   |                   ^^^

error: conflicting: both `bits` and `bytes` specified on field
  --> $DIR/bits_bytes_conflict.rs:24:28
   |
24 | struct Test4(#[deku(bits = "7", bytes = "8")] u8);
   |                            ^^^
